File: leetcode-practice-typescript/src/tree/208-implement-trie-prefix-tree.ts
class Trie {
    private readonly root: { [key: string]: any };

    constructor() {
        this.root = {};
    }

    insert(word: string): void {
        let node = this.root;
        for (const c of word) {
            if (!node[c]) {
                node[c] = {};
            }
            node = node[c];
        }
        node.isLeaf = true; // 标记为单词结尾
    }

    search(word: string): boolean {
        let node = this.root;
        for (const c of word) {
            if (!node[c]) return false;
            node = node[c];
        }
        return node.isLeaf === true;
    }

    startsWith(prefix: string): boolean {
        let node = this.root;
        for (const c of prefix) {
            if (!node[c]) return false;
            node = node[c];
        }
        return true;
    }
}

const trie = new Trie();
trie.insert("apple");
console.log(trie.search("apple"));   // 返回 True
console.log(trie.search("app"));     // 返回 False
console.log(trie.startsWith("app")); // 返回 True
trie.insert("app");
console.log(trie.search("app"));     // 返回 True

File: leetcode-practice-python/src/design/208_implement_trie_prefix_tree.py
class Trie:

    def __init__(self):
        self.children = [None] * 26
        self.isLeaf = False

    def insert(self, word: str) -> None:
        cur = self
        for c in word:
            index = ord(c) - ord("a")
            if not cur.children[index]:
                cur.children[index] = Trie()
            cur = cur.children[index]
        cur.isLeaf = True

    def search(self, word: str) -> bool:
        cur = self
        for c in word:
            index = ord(c) - ord("a")
            if not cur.children[index]:
                return False
            cur = cur.children[index]
        return cur.isLeaf

    def starts_with(self, prefix: str) -> bool:
        cur = self
        for c in prefix:
            index = ord(c) - ord("a")
            if not cur.children[index]:
                return False
            cur = cur.children[index]
        return True


if __name__ == "__main__":
    trie = Trie()
    trie.insert("apple")
    print(trie.search("apple"))
    print(trie.search("app"))
    print(trie.starts_with("app"))
